File: packages/shared/src/coding/lab-coding.ts
import type { ClinicalCode } from "./metric-coding.js";

const LOINC_SYSTEM = "http://loinc.org";

/**
 * Maps common lab test names to LOINC codes.
 * Used by the PDF extraction pipeline to auto-assign codes to extracted lab values.
 * Keys are normalized lowercase test names; the extraction pipeline should normalize
 * extracted test names before lookup.
 */
export const LAB_CODING: Record<string, ClinicalCode> = {
  // Lipid panel
  "total cholesterol": {
    code: "2093-3",
    display: "Cholesterol [Mass/volume] in Serum or Plasma",
    system: LOINC_SYSTEM,
  },
  cholesterol: {
    code: "2093-3",
    display: "Cholesterol [Mass/volume] in Serum or Plasma",
    system: LOINC_SYSTEM,
  },
  "hdl cholesterol": {
    code: "2085-9",
    display: "HDL Cholesterol",
    system: LOINC_SYSTEM,
  },
  hdl: {
    code: "2085-9",
    display: "HDL Cholesterol",
    system: LOINC_SYSTEM,
  },
  "ldl cholesterol": {
    code: "2089-1",
    display: "LDL Cholesterol",
    system: LOINC_SYSTEM,
  },
  ldl: {
    code: "2089-1",
    display: "LDL Cholesterol",
    system: LOINC_SYSTEM,
  },
  triglycerides: {
    code: "2571-8",
    display: "Triglycerides",
    system: LOINC_SYSTEM,
  },

  // Glucose / diabetes
  "hemoglobin a1c": {
    code: "4548-4",
    display: "Hemoglobin A1c/Hemoglobin.total in Blood",
    system: LOINC_SYSTEM,
  },
  hba1c: {
    code: "4548-4",
    display: "Hemoglobin A1c/Hemoglobin.total in Blood",
    system: LOINC_SYSTEM,
  },
  a1c: {
    code: "4548-4",
    display: "Hemoglobin A1c/Hemoglobin.total in Blood",
    system: LOINC_SYSTEM,
  },
  "fasting glucose": {
    code: "1558-6",
    display: "Fasting glucose",
    system: LOINC_SYSTEM,
  },
  glucose: {
    code: "1558-6",
    display: "Fasting glucose",
    system: LOINC_SYSTEM,
  },

  // Thyroid
  tsh: {
    code: "3016-3",
    display: "TSH",
    system: LOINC_SYSTEM,
  },
  "free t4": {
    code: "3024-7",
    display: "Free T4",
    system: LOINC_SYSTEM,
  },
  "free t3": {
    code: "3051-0",
    display: "Free T3",
    system: LOINC_SYSTEM,
  },

  // CBC
  hemoglobin: {
    code: "718-7",
    display: "Hemoglobin",
    system: LOINC_SYSTEM,
  },
  wbc: {
    code: "6690-2",
    display: "Leukocytes",
    system: LOINC_SYSTEM,
  },
  "white blood cells": {
    code: "6690-2",
    display: "Leukocytes",
    system: LOINC_SYSTEM,
  },
  rbc: {
    code: "789-8",
    display: "Erythrocytes",
    system: LOINC_SYSTEM,
  },
  "red blood cells": {
    code: "789-8",
    display: "Erythrocytes",
    system: LOINC_SYSTEM,
  },
  platelets: {
    code: "777-3",
    display: "Platelets",
    system: LOINC_SYSTEM,
  },

  // ── French lab report aliases ────────────────────────────────────────────
  // Haematology (CBC / Numération globulaire)
  hématies: { code: "789-8", display: "Erythrocytes", system: LOINC_SYSTEM },
  "globules rouges": { code: "789-8", display: "Erythrocytes", system: LOINC_SYSTEM },
  hémoglobine: { code: "718-7", display: "Hemoglobin", system: LOINC_SYSTEM },
  hematocrite: { code: "20570-8", display: "Hematocrit", system: LOINC_SYSTEM },
  hématocrite: { code: "20570-8", display: "Hematocrit", system: LOINC_SYSTEM },
  "v.g.m.": { code: "787-2", display: "MCV", system: LOINC_SYSTEM },
  vgm: { code: "787-2", display: "MCV", system: LOINC_SYSTEM },
  "t.c.m.h.": { code: "785-6", display: "MCH", system: LOINC_SYSTEM },
  tcmh: { code: "785-6", display: "MCH", system: LOINC_SYSTEM },
  "c.c.m.h.": { code: "786-4", display: "MCHC", system: LOINC_SYSTEM },
  ccmh: { code: "786-4", display: "MCHC", system: LOINC_SYSTEM },
  leucocytes: { code: "6690-2", display: "Leukocytes", system: LOINC_SYSTEM },
  "globules blancs": { code: "6690-2", display: "Leukocytes", system: LOINC_SYSTEM },
  plaquettes: { code: "777-3", display: "Platelets", system: LOINC_SYSTEM },
  "polynucléaires neutrophiles": {
    code: "751-8",
    display: "Neutrophils",
    system: LOINC_SYSTEM,
  },
  "polynucléaires éosinophiles": {
    code: "711-2",
    display: "Eosinophils",
    system: LOINC_SYSTEM,
  },
  "polynucléaires basophiles": {
    code: "704-7",
    display: "Basophils",
    system: LOINC_SYSTEM,
  },
  lymphocytes: { code: "731-0", display: "Lymphocytes", system: LOINC_SYSTEM },
  monocytes: { code: "742-7", display: "Monocytes", system: LOINC_SYSTEM },

  // Biochemistry (Biochimie sanguine)
  sodium: { code: "2951-2", display: "Sodium", system: LOINC_SYSTEM },
  potassium: { code: "2823-3", display: "Potassium", system: LOINC_SYSTEM },
  "glycemie a jeun": { code: "1558-6", display: "Fasting glucose", system: LOINC_SYSTEM },
  "glycémie à jeun": { code: "1558-6", display: "Fasting glucose", system: LOINC_SYSTEM },
  glycémie: { code: "1558-6", display: "Fasting glucose", system: LOINC_SYSTEM },
  uree: { code: "3094-0", display: "Urea nitrogen", system: LOINC_SYSTEM },
  urée: { code: "3094-0", display: "Urea nitrogen", system: LOINC_SYSTEM },
  creatinine: { code: "2160-0", display: "Creatinine", system: LOINC_SYSTEM },
  créatinine: { code: "2160-0", display: "Creatinine", system: LOINC_SYSTEM },
  "debit de filtration glomerulaire": {
    code: "33914-3",
    display: "eGFR (CKD-EPI)",
    system: LOINC_SYSTEM,
  },
  "débit de filtration glomérulaire": {
    code: "33914-3",
    display: "eGFR (CKD-EPI)",
    system: LOINC_SYSTEM,
  },
  crp: { code: "1988-5", display: "CRP", system: LOINC_SYSTEM },
  "proteines totales": { code: "2885-2", display: "Protein total", system: LOINC_SYSTEM },
  "protéines totales": { code: "2885-2", display: "Protein total", system: LOINC_SYSTEM },
  calcium: { code: "17861-6", display: "Calcium", system: LOINC_SYSTEM },
  ferritine: { code: "2276-4", display: "Ferritin", system: LOINC_SYSTEM },

  // Lipid panel (Bilan lipidique)
  "cholesterol total": { code: "2093-3", display: "Cholesterol total", system: LOINC_SYSTEM },
  "cholestérol total": { code: "2093-3", display: "Cholesterol total", system: LOINC_SYSTEM },
  "cholesterol hdl": { code: "2085-9", display: "HDL Cholesterol", system: LOINC_SYSTEM },
  "cholestérol hdl": { code: "2085-9", display: "HDL Cholesterol", system: LOINC_SYSTEM },
  "cholesterol ldl": { code: "2089-1", display: "LDL Cholesterol", system: LOINC_SYSTEM },
  "cholestérol ldl": { code: "2089-1", display: "LDL Cholesterol", system: LOINC_SYSTEM },
  triglycérides: { code: "2571-8", display: "Triglycerides", system: LOINC_SYSTEM },

  // Liver enzymes (Bilan hépatique)
  "asat - transaminases sgot": {
    code: "1920-8",
    display: "AST",
    system: LOINC_SYSTEM,
  },
  asat: { code: "1920-8", display: "AST", system: LOINC_SYSTEM },
  sgot: { code: "1920-8", display: "AST", system: LOINC_SYSTEM },
  "alat - transaminases sgpt": {
    code: "1742-6",
    display: "ALT",
    system: LOINC_SYSTEM,
  },
  alat: { code: "1742-6", display: "ALT", system: LOINC_SYSTEM },
  sgpt: { code: "1742-6", display: "ALT", system: LOINC_SYSTEM },
  "ggt - gamma-glutamyl-transferase": {
    code: "2324-2",
    display: "GGT",
    system: LOINC_SYSTEM,
  },
  ggt: { code: "2324-2", display: "GGT", system: LOINC_SYSTEM },
  "pal - phosphatases alcalines": {
    code: "6768-6",
    display: "Alkaline Phosphatase",
    system: LOINC_SYSTEM,
  },
  "phosphatases alcalines": {
    code: "6768-6",
    display: "Alkaline Phosphatase",
    system: LOINC_SYSTEM,
  },
  pal: { code: "6768-6", display: "Alkaline Phosphatase", system: LOINC_SYSTEM },
  "bilirubine totale": { code: "1975-2", display: "Bilirubin total", system: LOINC_SYSTEM },

  // Thyroid (Hormonologie)
  "tsh #": { code: "3016-3", display: "TSH", system: LOINC_SYSTEM },

  // Protein electrophoresis fractions
  albumine: { code: "1751-7", display: "Albumin", system: LOINC_SYSTEM },
};

File: apps/ios/HealthDashboard/Core/LabResults/LabCoding.swift
import Foundation

/// Maps normalised lab test name strings to LOINC codes.
///
/// Keys are lowercase, whitespace-trimmed test names (including common
/// abbreviations and French equivalents). The extraction pipeline normalises
/// extracted test names before lookup.
///
/// Mirrors `LAB_CODING` in `packages/shared/src/coding/lab-coding.ts`.
struct LoincEntry {
    let code: String
    let display: String
}

let labCoding: [String: LoincEntry] = {
    // English / universal
    var d: [String: LoincEntry] = [
        // Lipid panel
        "total cholesterol":    .init(code: "2093-3",  display: "Cholesterol [Mass/volume] in Serum or Plasma"),
        "cholesterol":          .init(code: "2093-3",  display: "Cholesterol [Mass/volume] in Serum or Plasma"),
        "hdl cholesterol":      .init(code: "2085-9",  display: "HDL Cholesterol"),
        "hdl":                  .init(code: "2085-9",  display: "HDL Cholesterol"),
        "ldl cholesterol":      .init(code: "2089-1",  display: "LDL Cholesterol"),
        "ldl":                  .init(code: "2089-1",  display: "LDL Cholesterol"),
        "triglycerides":        .init(code: "2571-8",  display: "Triglycerides"),

        // Glucose / diabetes
        "hemoglobin a1c":       .init(code: "4548-4",  display: "Hemoglobin A1c/Hemoglobin.total in Blood"),
        "hba1c":                .init(code: "4548-4",  display: "Hemoglobin A1c/Hemoglobin.total in Blood"),
        "a1c":                  .init(code: "4548-4",  display: "Hemoglobin A1c/Hemoglobin.total in Blood"),
        "fasting glucose":      .init(code: "1558-6",  display: "Fasting glucose"),
        "glucose":              .init(code: "1558-6",  display: "Fasting glucose"),

        // Thyroid
        "tsh":                  .init(code: "3016-3",  display: "TSH"),
        "tsh #":                .init(code: "3016-3",  display: "TSH"),
        "free t4":              .init(code: "3024-7",  display: "Free T4"),
        "free t3":              .init(code: "3051-0",  display: "Free T3"),

        // CBC
        "hemoglobin":           .init(code: "718-7",   display: "Hemoglobin"),
        "wbc":                  .init(code: "6690-2",  display: "Leukocytes"),
        "white blood cells":    .init(code: "6690-2",  display: "Leukocytes"),
        "rbc":                  .init(code: "789-8",   display: "Erythrocytes"),
        "red blood cells":      .init(code: "789-8",   display: "Erythrocytes"),
        "platelets":            .init(code: "777-3",   display: "Platelets"),
        "hematocrit":           .init(code: "20570-8", display: "Hematocrit"),
        "mcv":                  .init(code: "787-2",   display: "MCV"),
        "mch":                  .init(code: "785-6",   display: "MCH"),
        "mchc":                 .init(code: "786-4",   display: "MCHC"),
        "neutrophils":          .init(code: "751-8",   display: "Neutrophils"),
        "lymphocytes":          .init(code: "731-0",   display: "Lymphocytes"),
        "monocytes":            .init(code: "742-7",   display: "Monocytes"),
        "eosinophils":          .init(code: "711-2",   display: "Eosinophils"),
        "basophils":            .init(code: "704-7",   display: "Basophils"),

        // Metabolic / renal
        "sodium":               .init(code: "2951-2",  display: "Sodium"),
        "potassium":            .init(code: "2823-3",  display: "Potassium"),
        "creatinine":           .init(code: "2160-0",  display: "Creatinine"),
        "urea":                 .init(code: "3094-0",  display: "Urea nitrogen"),
        "bun":                  .init(code: "3094-0",  display: "Urea nitrogen"),
        "egfr":                 .init(code: "33914-3", display: "eGFR (CKD-EPI)"),
        "calcium":              .init(code: "17861-6", display: "Calcium"),
        "crp":                  .init(code: "1988-5",  display: "CRP"),
        "protein total":        .init(code: "2885-2",  display: "Protein total"),
        "albumin":              .init(code: "1751-7",  display: "Albumin"),
        "ferritin":             .init(code: "2276-4",  display: "Ferritin"),

        // Liver
        "ast":                  .init(code: "1920-8",  display: "AST"),
        "sgot":                 .init(code: "1920-8",  display: "AST"),
        "alt":                  .init(code: "1742-6",  display: "ALT"),
        "sgpt":                 .init(code: "1742-6",  display: "ALT"),
        "ggt":                  .init(code: "2324-2",  display: "GGT"),
        "alkaline phosphatase": .init(code: "6768-6",  display: "Alkaline Phosphatase"),
        "alp":                  .init(code: "6768-6",  display: "Alkaline Phosphatase"),
        "bilirubin":            .init(code: "1975-2",  display: "Bilirubin total"),
        "total bilirubin":      .init(code: "1975-2",  display: "Bilirubin total"),

        // ── French aliases ───────────────────────────────────────────────────
        // Haematologie
        "hématies":                       .init(code: "789-8",   display: "Erythrocytes"),
        "globules rouges":                .init(code: "789-8",   display: "Erythrocytes"),
        "hémoglobine":                    .init(code: "718-7",   display: "Hemoglobin"),
        "hematocrite":                    .init(code: "20570-8", display: "Hematocrit"),
        "hématocrite":                    .init(code: "20570-8", display: "Hematocrit"),
        "v.g.m.":                         .init(code: "787-2",   display: "MCV"),
        "vgm":                            .init(code: "787-2",   display: "MCV"),
        "t.c.m.h.":                       .init(code: "785-6",   display: "MCH"),
        "tcmh":                           .init(code: "785-6",   display: "MCH"),
        "c.c.m.h.":                       .init(code: "786-4",   display: "MCHC"),
        "ccmh":                           .init(code: "786-4",   display: "MCHC"),
        "leucocytes":                     .init(code: "6690-2",  display: "Leukocytes"),
        "globules blancs":                .init(code: "6690-2",  display: "Leukocytes"),
        "plaquettes":                     .init(code: "777-3",   display: "Platelets"),
        "polynucléaires neutrophiles":    .init(code: "751-8",   display: "Neutrophils"),
        "polynucléaires éosinophiles":    .init(code: "711-2",   display: "Eosinophils"),
        "polynucléaires basophiles":      .init(code: "704-7",   display: "Basophils"),

        // Biochimie
        "glycemie a jeun":                .init(code: "1558-6",  display: "Fasting glucose"),
        "glycémie à jeun":                .init(code: "1558-6",  display: "Fasting glucose"),
        "glycémie":                       .init(code: "1558-6",  display: "Fasting glucose"),
        "uree":                           .init(code: "3094-0",  display: "Urea nitrogen"),
        "urée":                           .init(code: "3094-0",  display: "Urea nitrogen"),
        "créatinine":                     .init(code: "2160-0",  display: "Creatinine"),
        "debit de filtration glomerulaire": .init(code: "33914-3", display: "eGFR (CKD-EPI)"),
        "débit de filtration glomérulaire": .init(code: "33914-3", display: "eGFR (CKD-EPI)"),
        "proteines totales":              .init(code: "2885-2",  display: "Protein total"),
        "protéines totales":              .init(code: "2885-2",  display: "Protein total"),
        "ferritine":                      .init(code: "2276-4",  display: "Ferritin"),

        // Bilan lipidique
        "cholesterol total":              .init(code: "2093-3",  display: "Cholesterol total"),
        "cholestérol total":              .init(code: "2093-3",  display: "Cholesterol total"),
        "cholesterol hdl":                .init(code: "2085-9",  display: "HDL Cholesterol"),
        "cholestérol hdl":                .init(code: "2085-9",  display: "HDL Cholesterol"),
        "cholesterol ldl":                .init(code: "2089-1",  display: "LDL Cholesterol"),
        "cholestérol ldl":                .init(code: "2089-1",  display: "LDL Cholesterol"),
        "triglycérides":                  .init(code: "2571-8",  display: "Triglycerides"),

        // Bilan hépatique
        "asat - transaminases sgot":      .init(code: "1920-8",  display: "AST"),
        "asat":                           .init(code: "1920-8",  display: "AST"),
        "alat - transaminases sgpt":      .init(code: "1742-6",  display: "ALT"),
        "alat":                           .init(code: "1742-6",  display: "ALT"),
        "ggt - gamma-glutamyl-transferase": .init(code: "2324-2", display: "GGT"),
        "pal - phosphatases alcalines":   .init(code: "6768-6",  display: "Alkaline Phosphatase"),
        "phosphatases alcalines":         .init(code: "6768-6",  display: "Alkaline Phosphatase"),
        "pal":                            .init(code: "6768-6",  display: "Alkaline Phosphatase"),
        "bilirubine totale":              .init(code: "1975-2",  display: "Bilirubin total"),

        // Hormonologie
        "albumine":                       .init(code: "1751-7",  display: "Albumin"),
    ]
    return d
}()

/// Looks up the LOINC entry for a raw test name.
///
/// Normalises to lowercase and trims whitespace before lookup.
/// Returns `nil` if no mapping is found.
func lookupLoincCode(for rawName: String) -> LoincEntry? {
    labCoding[rawName.lowercased().trimmingCharacters(in: .whitespaces)]
}
